block/block.go

```diff
@@ -1,374 +1,328 @@
+func (kind Kind) String() string {
+	switch kind {
+	case Standard:
+		return "standard"
+	case Rebase:
+		return "rebase"
+	case Base:
+		return "base"
+	default:
+		panic(fmt.Errorf("invariant violation: unexpected kind=%d", uint8(kind)))
 	}
-	txHeaders := make([]byte, 32*len(block.Txs))
-	for i, tx := range block.Txs {
-		copy(txHeaders[32*i:], tx[:])
-	}
-	block.TxHeader = sha3.Sum256(txHeaders)
-	block.Header = sha3.Sum256([]byte(block.String()))
-	return block
 }
 
-// Equal checks for equality between the heights,  headers, and parent headers.
-func (block Block) Equal(other Block) bool {
-	return block.Height == other.Height &&
-		block.Header.Equal(other.Header) &&
-		block.ParentHeader.Equal(other.ParentHeader)
+// A Header defines properties of a Block that are not application-specific.
+// These properties are required by, or produced by, the consensus algorithm.
+type Header struct {
+	kind       Kind      // Kind of Block
+	parentHash id.Hash   // Hash of the Block parent
+	baseHash   id.Hash   // Hash of the Block base
+	height     Height    // Height at which the Block was committed
+	round      Round     // Round at which the Block was committed
+	timestamp  Timestamp // Seconds since Unix Epoch
+
+	// Signatories oversee the consensus algorithm (must be nil unless the Block
+	// is a Rebase Block)
+	signatories id.Signatories
 }
 
-// Sign the header of the Block.
-func (block Block) Sign(signer sig.Signer) (SignedBlock, error) {
-	signedBlock := SignedBlock{
-		Block: block,
-	}
-
-	signature, err := signer.Sign(signedBlock.Header)
-	if err != nil {
-		return SignedBlock{}, err
+// NewHeader returns a Header. It will panic a pre-condition for Header validity
+// is violated.
+func NewHeader(kind Kind, parentHash, baseHash id.Hash, height Height, round Round, timestamp Timestamp, signatories id.Signatories) Header {
+	switch kind {
+	case Standard:
+		if signatories != nil {
+			panic("pre-condition violation: standard blocks must not declare signatories")
+		}
+	case Rebase:
+		if signatories == nil || len(signatories) == 0 {
+			panic("pre-condition violation: rebase blocks must declare signatories")
+		}
+	case Base:
+		if signatories == nil || len(signatories) == 0 {
+			panic("pre-condition violation: base blocks must declare signatories")
+		}
+	default:
+		panic(fmt.Errorf("pre-condition violation: unexpected block kind=%v", kind))
 	}
-	signedBlock.Signature = signature
-	signedBlock.Signatory = signer.Signatory()
-
-	return signedBlock, nil
-}
-
-// String implements the Stringer interface for the Block type.
-func (block Block) String() string {
-	return fmt.Sprintf("Block(Height=%d,Timestamp=%d,TxHeader=%s,ParentHeader=%s)", block.Height, block.Time.Unix(), block.TxHeader, block.ParentHeader)
-}
-
-func (block Block) Write(w io.Writer) error {
-	if err := binary.Write(w, binary.LittleEndian, block.Time.Unix()); err != nil {
-		return err
+	if parentHash.Equal(InvalidHash) {
```

Comment:
The programmer is expected to define some kind of globally agreed upon "genesis" block. This is needed because it is up to the programmer to define the genesis signatories and state.